# HG changeset patch
# User Alexandre Lissy <lissyx+mozillians@lissyx.dyndns.org>
# Date 1787307304 -7200
# Node ID 594eb348a49a2ec1b20c357fc33c0bb16a1052dc
# Parent  c1beed3d0e89c961441967dae98868182ff84f2e
Bug 2007074 - Handle XDG_CONFIG_HOME for profile on Thunderbird r=thunderbird-back-end-reviewers,Thinker,babolivier

Enable handling of XDG on Thunderbird, and update gtest accordingly.

Differential Revision: https://phabricator.services.mozilla.com/D309109

diff --git a/toolkit/xre/nsXREDirProvider.cpp b/toolkit/xre/nsXREDirProvider.cpp
--- a/toolkit/xre/nsXREDirProvider.cpp
+++ b/toolkit/xre/nsXREDirProvider.cpp
@@ -1342,9 +1342,14 @@ nsresult nsXREDirProvider::AppendFromApp
   // TODO: Bug 1990407 - Evaluate if refactoring might be required there in the
   // future?
   if (gAppData->profile) {
     nsAutoCString profile;
-    profile = gAppData->profile;
+#  if defined(MOZ_THUNDERBIRD)
+    if (gAppData->profile[0] != '.') {
+      profile.Assign('.');
+    }
+#  endif
+    profile.Append(gAppData->profile);
     MOZ_TRY(aFile->AppendRelativeNativePath(profile));
   } else {
     nsAutoCString vendor;
     nsAutoCString appName;
@@ -1363,9 +1368,10 @@ nsresult nsXREDirProvider::AppendFromApp
 
 /*
  * Check if legacy directory exists, which can be:
  *  (1) $HOME/.<gAppData->vendor>/<gAppData->appName>
- *  (2) $HOME/<gAppData->profile>
+ *  (2) $HOME/<gAppData->profile> on Firefox or
+ *      $HOME/.<gAppData->profile> on Thunderbird
  *  (3) $HOME/<MOZ_USER_DIR>
  *
  * The MOZ_USER_DIR will also be defined in case (1), so first check the deeper
  * directory.
@@ -1521,9 +1527,22 @@ nsresult nsXREDirProvider::GetLegacyOrXD
     }
 
     // If the build was made against a specific profile name, MOZ_APP_PROFILE=
     // then make sure we respect this and dont move to XDG directory
-    if (gAppData->profile) {
+    //
+    // Only do this on Firefox.
+    //
+    // For Thunderbird, it always defines MOZ_APP_PROFILE=thunderbird. Hence,
+    // check if the gAppData->profile value is set to something else, indicating
+    // of a specific build where legacy needs to be forced, similar to the
+    // Firefox handling. Both casing are verified to account for systems where
+    // that matters.
+    if (gAppData->profile
+#  if defined(MOZ_THUNDERBIRD)
+        && strcmp(gAppData->profile, "thunderbird") != 0 &&
+        strcmp(gAppData->profile, "Thunderbird") != 0
+#  endif
+    ) {
       MOZ_TRY(NS_NewNativeLocalFile(nsDependentCString(aHomeDir),
                                     getter_AddRefs(localDir)));
     } else {
       MOZ_TRY(GetLegacyOrXDGConfigHome(aHomeDir, getter_AddRefs(localDir)));
